File: VB/Behavioural/Visitor/VisitorPatternRunner.vb
Public Class VisitorPatternRunner
	Public Sub Run()
		Dim employeeJJ = New SalesEmployee("Justus Jonas", 75000, 25000)
		Dim employeePS = New SalesEmployee("Peter Shaw", 60199, 12500)
		Dim employeeBA = New BackOfficeEmployee("Bob Andrews", 57600, 10750)

		Dim specialEmployees As New List(Of IVisitable) From {
			employeeJJ,
			employeePS,
			employeeBA
		}

		For Each specialEmployee In specialEmployees
			Dim compensationCalculator = New CalculateTotalCompensationVisitor
			specialEmployee.Accept(compensationCalculator)
			Dim baseEmployee = TryCast(specialEmployee, Employee)

			If baseEmployee IsNot Nothing Then
				Console.WriteLine($"Total Compensation For {baseEmployee.Name} : {compensationCalculator.TotalCompensation} EUR")
			End If
		Next

	End Sub

	Friend Interface IVisitor
		Sub Visit(employee As BackOfficeEmployee)
		Sub Visit(employee As SalesEmployee)
	End Interface
	Friend Interface IVisitable
		Sub Accept(visitor As IVisitor)
	End Interface
	Friend Class Employee
		Public Name As String
		Public Salary As Decimal

		Public Sub New(name As String, salary As Decimal)
			Me.Name = name
			Me.Salary = salary
		End Sub
	End Class


	Friend Class BackOfficeEmployee : Inherits Employee
		Implements IVisitable

		Friend Bonus As Decimal = 12500
		Public Sub New(name As String, salary As Decimal, bonus As Decimal)
			MyBase.New(name, salary)
			Me.Bonus = bonus
		End Sub

		Public Sub Accept(visitor As IVisitor) Implements IVisitable.Accept
			visitor.Visit(Me)
		End Sub
	End Class
	Friend Class SalesEmployee : Inherits Employee
		Implements IVisitable

		Friend Commissions As Decimal = 12500
		Public Sub New(name As String, salary As Decimal, commissions As Decimal)
			MyBase.New(name, salary)
			Me.Commissions = commissions
		End Sub

		Public Sub Accept(visitor As IVisitor) Implements IVisitable.Accept
			visitor.Visit(Me)
		End Sub
	End Class

	Friend Class CalculateTotalCompensationVisitor
		Implements IVisitor

		Private _totalCompensation As Decimal
		Friend ReadOnly Property TotalCompensation() As Decimal
			Get
				Return _totalCompensation
			End Get
		End Property

		Public Sub Visit(employee As BackOfficeEmployee) Implements IVisitor.Visit
			_totalCompensation = employee.Salary + employee.Bonus
		End Sub

		Public Sub Visit(employee As SalesEmployee) Implements IVisitor.Visit
			_totalCompensation = employee.Salary + employee.Commissions
		End Sub
	End Class

End Class

File: CS/Behavioural/Visitor/VisitorPatternRunner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pattern.Behavioural.Visitor
{
	internal static class VisitorPatternRunner
	{
		public static void Run()
		{
			Console.WriteLine("Visitor Pattern Runner ..");

			var employeeJJ = new SalesEmployee("Justus Jonas", 100000, 27500);
			var employeePS = new SalesEmployee("Peter Shaw", 70000, 23250);
			var employeeBA = new BackOfficeEmployee("Bob Andrews", 75000, 21275);
			
			var specialEmployeeList = new List<IVisitableElement>(){ employeeJJ, employeePS, employeeBA };
			
			var totalCompensationVisitor = new CompensationVisitor();
			
			foreach (var specialEmployee in specialEmployeeList) 
			{
				specialEmployee.Accept(totalCompensationVisitor);
				
				var baseEmployee = (Employee) specialEmployee;
				if(baseEmployee!= null)
				{
					var specialEmployeeCompensationVisitor = new CompensationVisitor();
					specialEmployee.Accept(specialEmployeeCompensationVisitor);
					Console.WriteLine("Compensation for " + baseEmployee.Name + " is " + specialEmployeeCompensationVisitor.TotalCompensation);
				}				
			}

			Console.WriteLine("Total Compensation for " + employeeJJ.Name + ", " + employeePS.Name + ", " + employeeBA.Name + " is " + totalCompensationVisitor.TotalCompensation);
		}

		public interface IVisitableElement
		{
			void Accept(IVisitor visitor);
		}

		public interface IVisitor 
		{
			void Visit(BackOfficeEmployee e);
			void Visit(SalesEmployee e);
		}

		public class CompensationVisitor : IVisitor
		{
			public decimal TotalCompensation { get; private set; } = 0;
			public void Visit(BackOfficeEmployee e)
			{
				TotalCompensation += e.Salary + e.Bonus;
			}

			public void Visit(SalesEmployee e)
			{
				TotalCompensation += e.Salary + e.Commission;
			}
		}
		public class Employee
		{
			public string Name { get; set; }
			public decimal Salary { get; set; }

			public Employee(string name, decimal salary) 
			{
				Name = name;
				Salary = salary;
			}
		}

		public class BackOfficeEmployee : Employee, IVisitableElement
		{
			public decimal Bonus { get; set; }

			public BackOfficeEmployee(string name, decimal salary, decimal bonus) : base(name, salary)
			{
				Bonus = bonus;
			}

			public void Accept(IVisitor visitor)
			{
				visitor.Visit(this);
			}
		}
		public class SalesEmployee : Employee, IVisitableElement
		{
			public decimal Commission { get; set; }

			public SalesEmployee(string name, decimal salary, decimal comission) : base(name, salary)
			{
				Commission = comission;
			}
			public void Accept(IVisitor visitor)
			{
				visitor.Visit(this);
			}
		}
	}
}
